### PR TITLE
Implementation of smoothness encoded value.

### PR DESCRIPTION
**Pull request by ratrun**
This is the implementation of the encoded value for the `smoothness`. With commit https://github.com/graphhopper/graphhopper/pull/2303/commits/68932aa7551a97825eafdcabc945d884cfcd9943 the bicycle common flag encoder now use it.

implements #655, supersedes #2108

### CODE REVIEW

**Comment by ratrun** [comment]
I believe that now the code is working as you suggested in #2108. 
Anyway, I had two problems with failing tests:

1.  The first one is in AbstractBikeFlagEncoderTester, see the changes I had to make. This test looks a bit ugly to me. I think the flagencoders are using more bits now, such that the encoded value does not fit into one integer any longer, so now I'm oring the values of the two 32 bit integeres.
2. The second one are two failing tests in PrepareContractionHierarchiesTest. I have no idea why these two are failing now.

**Comment by easbar** [comment]
> The second one are two failing tests in PrepareContractionHierarchiesTest. I have no idea why these two are failing now.

I think these tests were somewhat broken already before and only worked 'accidentally'. Please update your branch to 
https://github.com/graphhopper/graphhopper/commit/b18d576f6c125a6db8fe99e8ec4712c4a31f4598 or newer and the problem should be solved.

**Comment by ratrun** [comment]
@easbar Thank you for the fix, I can confirm that it is working again. I have now re-enabled these tests.

**Review by karussell — CHANGES_REQUESTED** [review]
Thank you - this looks good! Can you have a look into my questions?

File: core/src/test/java/com/graphhopper/routing/ch/PrepareContractionHierarchiesTest.java [hunk 1]
```diff
@@ -571,7 +571,7 @@ private void checkPath(GraphHopperStorage g, CHConfig c, int expShortcuts, doubl
         PrepareContractionHierarchies prepare = createPrepareContractionHierarchies(g, c);
         useNodeOrdering(prepare, nodeOrdering);
         prepare.doWork();
-        assertEquals(expShortcuts, prepare.getShortcuts(), c.toString());
+        //assertEquals(expShortcuts, prepare.getShortcuts(), c.toString());
```
[the remark this answers is not in the document]
Comment (ratrun) [reply core/src/test/java/com/graphhopper/routing/ch/PrepareContractionHierarchiesTest.java:574]: Sorry, this was still commented per accident. Done now!

File: core/src/test/java/com/graphhopper/routing/util/AbstractBikeFlagEncoderTester.java [hunk 2]
```diff
@@ -255,21 +255,21 @@ public void testTramStations() {
...
+        assertNotEquals(0, flags.ints[0] | flags.ints[1]);
 
         way = new ReaderWay(1);
         way.setTag("highway", "track");
         way.setTag("railway", "platform");
         way.setTag("bicycle", "no");
 
         flags = encoder.handleWayTags(encodingManager.createEdgeFlags(), way, encoder.getAccess(way));
-        assertEquals(0, flags.ints[0]);
+        assertEquals(0, flags.ints[0] | flags.ints[1]);
```
[the remark this answers is not in the document]
Comment (ratrun) [reply core/src/test/java/com/graphhopper/routing/util/AbstractBikeFlagEncoderTester.java:272]: Thank you, Done.

**Review by karussell — APPROVED** [review]

File: core/src/main/java/com/graphhopper/routing/util/MountainBikeFlagEncoder.java [hunk 3]
```diff
@@ -128,6 +128,16 @@ protected MountainBikeFlagEncoder(int speedBits, double speedFactor, int maxTurn
...
 
+        setSmoothnessSpeedFactor(com.graphhopper.routing.ev.Smoothness.EXCELLENT, 1.1d);
+        setSmoothnessSpeedFactor(com.graphhopper.routing.ev.Smoothness.GOOD, 1.0d);
+        setSmoothnessSpeedFactor(com.graphhopper.routing.ev.Smoothness.INTERMEDIATE, 0.9d);
+        setSmoothnessSpeedFactor(com.graphhopper.routing.ev.Smoothness.BAD, 0.7d);
+        setSmoothnessSpeedFactor(com.graphhopper.routing.ev.Smoothness.VERY_BAD, 0.6d);
+        setSmoothnessSpeedFactor(com.graphhopper.routing.ev.Smoothness.HORRIBLE, 0.5d);
+        setSmoothnessSpeedFactor(com.graphhopper.routing.ev.Smoothness.VERY_HORRIBLE, 0.4d);
+        // SmoothnessSpeed <= smoothnessFactorPushingSectionThreshold gets mapped to speed PUSHING_SECTION_SPEED
+        setSmoothnessSpeedFactor(com.graphhopper.routing.ev.Smoothness.IMPASSABLE, smoothnessFactorPushingSectionThreshold);
```
Comment (karussell) [inline core/src/main/java/com/graphhopper/routing/util/MountainBikeFlagEncoder.java:139]: One last question: do we need to overwrite this here if already set in BikeFlagEncoder?

File: core/src/test/java/com/graphhopper/routing/util/BikeFlagEncoderTest.java [hunk 4]
```diff
@@ -290,6 +290,54 @@ public void testSpeedAndPriority() {
...
     }
 
+    @Test
+    public void testSmoothness() {
+        ReaderWay way = new ReaderWay(1);
+        way.setTag("highway", "residential");
+        way.setTag("smoothness", "excellent");
+        IntsRef relFlags = encodingManager.createRelationFlags();
+        IntsRef flags = encodingManager.handleWayTags(way, accessMap, relFlags);
+        assertEquals(20, encoder.getAverageSpeedEnc().getDecimal(false, flags), 0.01);
```
Comment (karussell) [inline core/src/test/java/com/graphhopper/routing/util/BikeFlagEncoderTest.java:300]: Can you replace this (and the other asserts) with getSpeedFromFlags?

File: core/src/main/java/com/graphhopper/routing/util/MountainBikeFlagEncoder.java (hunk 3 above)
Comment (ratrun) [reply core/src/main/java/com/graphhopper/routing/util/MountainBikeFlagEncoder.java:139]: As MountainBikeFlagEncoder extends BikeCommonFlagEncoder and not BikeFlagEncoder I would say yes!

File: core/src/test/java/com/graphhopper/routing/util/BikeFlagEncoderTest.java (hunk 4 above)
Comment (ratrun) [reply core/src/test/java/com/graphhopper/routing/util/BikeFlagEncoderTest.java]: Done in 26f4da7.

File: core/src/main/java/com/graphhopper/routing/util/MountainBikeFlagEncoder.java (hunk 3 above)
Comment (karussell) [reply core/src/main/java/com/graphhopper/routing/util/MountainBikeFlagEncoder.java:139]: Ah, ok :+1:

File: core/src/test/java/com/graphhopper/routing/util/AbstractBikeFlagEncoderTester.java [hunk 5]
```diff
@@ -70,7 +70,8 @@ protected IntsRef assertPriority(int expectedPrio, ReaderWay way, ReaderRelation
     }
 
     protected double getSpeedFromFlags(ReaderWay way) {
-        IntsRef flags = encoder.handleWayTags(encodingManager.createEdgeFlags(), way, WAY);
+        IntsRef relFlags = encodingManager.createRelationFlags();
+        IntsRef flags = encodingManager.handleWayTags(way, new EncodingManager.AcceptWay().put(encoder.toString(), WAY), relFlags);
         return avgSpeedEnc.getDecimal(false, flags);
```
Comment (karussell) [inline core/src/test/java/com/graphhopper/routing/util/AbstractBikeFlagEncoderTester.java:75]: Ups, this looked so identical - thanks for fixing this!

**Comment by karussell** [comment]
Thanks for your patience and work to make this merging at the end :) !

**Pull request merged**